x/lazyledgerapp/client/builder.go

```diff
@@ -0,0 +1,172 @@
+package client
+
+import (
+	"context"
+
+	sdkclient "github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/lazyledger/lazyledger-app/app/params"
+	"google.golang.org/grpc"
+)
+
+// Builder wraps the standard sdkclient.TxBuilder add a few quality of life methods along
+// with keeping track of the account number to produce transactions
+type Builder struct {
+	sdkclient.TxBuilder
+	encCfg               params.EncodingConfig
+	address              sdktypes.Address
+	chainID              string
+	accountNum, sequence uint64
+}
+
+func NewBuilder(address sdktypes.Address, chainID string) *Builder {
+	encCfg := params.RegisterAccountInterface(params.MakeEncodingConfig())
+	txBuilder := encCfg.TxConfig.NewTxBuilder()
+	return &Builder{
+		TxBuilder: txBuilder,
+		encCfg:    encCfg,
+		address:   address,
+		chainID:   chainID,
+	}
+}
+
+// UpdateAccountNumber queries the applicaiton to find the latest account number and
+// sequence, updating the respective internal fields
+func (b *Builder) UpdateAccountNumber(ctx context.Context, conn *grpc.ClientConn) error {
+	accNum, seqNumb, err := QueryAccount(ctx, conn, b.encCfg, b.address.String())
+	if err != nil {
+		return err
+	}
+	b.accountNum = accNum
+	b.sequence = seqNumb
+	return nil
+}
+
+// BuildSignedTx creates a signed sdk.Tx that contains the provided WirePayForMessage.
+// The internal fees, the account number, and the sequence should be updated before calling this function
+// (see b.UpdateAccountNumber, sdkclient.TxBuilder.SetGasLimit, and sdkclient.TxBuilder.SetFeeAmount)
+func (b *Builder) BuildSignedTx(msg sdktypes.Msg, kr keyring.Keyring) (authsigning.Tx, error) {
```

Comment:
Two questions:
1. Is this only supposed to be used with WirePayForMessage (if yes: should we name this BuildWirePayForMessage or sth like that?)
2. related to the above: if this is only supposed to be used with a WirePayForMessage, should this take a WirePayForMessage as param instead of a more general sdktypes.Msg? (if yes, I think the name could stay the same as the types would more clearly communicate what this is building & what the diff between this and vanilla TxBuilder is).

Reply:
>Is this only supposed to be used with WirePayForMessage (if yes: should we name this BuildWirePayForMessage or sth like that?)

Since this code basically handles the [programmatic signing of transactions](https://docs.cosmos.network/v0.42/run-node/txs.html#signing-a-transaction-2), I figured that it didn't have to be specific to `WirePayForMessage`. I think that this could be useful in the future as it allows users to use the same signing code for simple token transfers, staking txs, etc.

I think a great example of what this actually looks like is in the [optimint PR](https://github.com/lazyledger/lazyledger-app/blob/683ebe4e579b9795dbdb3da1dd1ee258c897c386/x/lazyledgerapp/client/builder.go#L53)

Does that also answer the second question?